### ISSUE TITLE
Pseudomods and Compute.cs

### ISSUE THREAD

**Issue by MLanghof** [brather1ng/PoESkillTree#3]
I think there is no doubt that nobody wants to split his increased damage in constraints on (using fireball as example) "#% increased projectile damage", "#% increased elemental damage", "#% increased spell damage", "#% increased fire damage" and so on.

One solution would be that users can create some sort of pseudo-mods, specifying which increases get added into it. In this example you could call it "Fireball increased damage" and throw in all the relevant mods.

We do have Compute.cs with a fully-fledged damage calculation system in place though, so that you would (theoretically) only have so specify you're interested in fireball dps, the code can then figure out all appropriate modifiers. However, even this gets quickly complicated by skills having multiple damaging parts (fireball hit <-> aoe explosion (<-> burn <-> burn prolif?) are affected by different increases).

In any case, we need an abstraction layer above the tree attributes. Potentially a recursive structure, as in an abstract `Attribute` class, inherited by `PseudoAttribute` and `TreeAttribute`, with the former storing a set of `Attribute`s. Unless we use something like a tree view for that, it's going to be rough to communicate it in the UI though.

In effect I would think most people will be interested in a dps figure, a life figure and some other defensive figures. Maybe this is already going quite far into the direction of the Automated tab, I don't know. An idea would be to make the Automated tab nothing else than a preset for the Advanced tab, so that you a) could dive into the Advanced tab with something to start from and b) save your Advanced tab setting as an automated tab.

Seeing as total dps/life etc. figures can only really be attained when your gear is known, it might make sense to draw the line there: A solver for combined gear+tree (you can optimize based on dps, life etc.) and a solver for purely the tree (using pseudomods?). It would be awesome if this could match 1:1 on the Automated <-> Advanced tabs, but I'm not convinced that covers all use cases.

Ok, enough rambling from my end. Curious to hear what you think about this.

**Comment by MLanghof** [comment]
Now that I've written down my train of though, things are becoming much clearer.

It really makes no sense to multiply any stats originating from purely the tree, since almost all will have further increases on gear (read: To multiply anything, you need to make assumptions about gear. Or is there a counterexample?). Therefore, the tree-only solver should have a single layer of pseudo-attributes. Each pseudo-attribute has a list of tree attributes that play into it, and you only weight the pseudo-attributes. This would even enable proper handling of the fireball hit vs aoe stuff (again, within the limits of not knowing the gear).

Compute.cs can wait for when item imports are included in some shape or form I suppose.

**Comment by brather1ng** [comment]
Splitting it into tree only and tree+gear and having a single layer of pseudo attributes sounds good. 

For the Advanced Solver this could be done by having a checkbox that, if it's set to tree only, has settings that allow you to specify on what the pseudo attributes are based on (like the "Pseudo Section Settings" in the POEMate App seen in the video https://www.youtube.com/watch?v=G8QM16bFLbE&feature=youtu.be, I only have seen that from the video though). If it's set to tree+gear constraints like dps, life, etc. are enabled.
In both cases the raw attributes, like they are now, and the pseudo attributes that combine them are enabled.

For the Automated Solver this would basically be same, either include gear or give the tool the settings needed to decide what to optimize. Having it only as a preset for the Advanced Solver sounds nice, but it would have the need for the program to now what target values it would have. That might include too much hardcoding of values for each different build. Imo the best solution would be if it doesn't have target values and more like weighting constraints against each other. But discussing this further can be wait until the Advanced Solver is in a proper state, I guess.

Also I'm not sure that Compute.cs is not needed for the pseudo attributes. It can get somewhat complicated if we want to include each edge case with all the keystones affecting what is included in the attributes (strength only influences increased physical damage for a bow build if Iron Grip is taken and stuff like that).

**Comment by MLanghof** [comment]
Yeah, we can ignore Compute.cs for the tree-only pseudo-attribute stuff. I think that's a good second milestone (the first being what you've already achieved).

**Comment by brather1ng** [comment]
I have finished Pseudo Attribute Constraints now how I imagined them. See "Pseudo Attributes" and "Settings" tabs in AdvancedSolver. Definitions for them are loaded from xml files in [WPFSKillTree/Data/PseudoAttribute](https://github.com/brather1ng/PoESkillTree/tree/treeGen/WPFSKillTree/Data/PseudoAttributes).

If the user wants to edit xml stuff it would technically also support allowing the user to write his own pseudo attributes (which is pretty easy if you ignore conditions), though with the conditions pretty much every combination that is not too far off can be included in like 20 pre defined ones.
